numba/targets/linalg.py

```diff
@@ -1001,20 +1013,68 @@ def qr_impl(a):
 
 
 # helpers and jitted specialisations required for np.linalg.lstsq
-def _check_linalg_1_or_2d_matrix(a, func_name):
-    # checks that a matrix is 1 or 2D
-    if not isinstance(a, types.Array):
-        raise TypingError("np.linalg.%s() only supported for array types "
-                          % func_name)
-    if not a.ndim <= 2:
-        raise TypingError("np.linalg.%s() only supported on 1 and 2-D arrays "
-                          % func_name)
-    if not isinstance(a.dtype, (types.Float, types.Complex)):
-        raise TypingError("np.linalg.%s() only supported on "
-                          "float and complex arrays." % func_name)
+# and np.linalg.solve. These functions have "system" in their name
+# as a differentiator.
 
+def _get_system_copy_in_b_impl(b):
+    # gets an implementation for correctly copying 'b' into the
+    # scratch space for 'b'
+    ndim = b.ndim
+    if ndim == 1:
+        @jit(nopython=True)
+        def oneD_impl(bcpy, b, nrhs):
+            bcpy[:b.shape[-1], 0] = b
+        return oneD_impl
+    else:
+        @jit(nopython=True)
+        def twoD_impl(bcpy, b, nrhs):
+            bcpy[:b.shape[-2], :nrhs] = b
+        return twoD_impl
+
+
+def _get_system_compute_nrhs(b):
+    # gets and implementation for computing the number of right hand
+    # sides in the system of equations
+    ndim = b.ndim
+    if ndim == 1:
+        @jit(nopython=True)
+        def oneD_impl(b):
+            return 1
+        return oneD_impl
+    else:
+        @jit(nopython=True)
+        def twoD_impl(b):
+            return b.shape[-1]
+        return twoD_impl
 
-def _get_res_impl(dtype, real_dtype, b):
+
+def _get_system_check_dimensionally_valid_impl(a, b):
+    # gets and implementation for checking that AX=B style system
+    # input is dimensionally valid
+    ndim = b.ndim
+    if ndim == 1:
+        @jit(nopython=True)
+        def oneD_impl(a, b):
+            am = a.shape[-2]
+            bm = b.shape[-1]
+            if am != bm:
+                raise np.linalg.LinAlgError(
+                    "Incompatible array sizes, system is not dimensionally valid.")
+        return oneD_impl
+    else:
+        @jit(nopython=True)
+        def twoD_impl(a, b):
+            am = a.shape[-2]
+            bm = b.shape[-2]
```

Comment:
... and so is this one, which means you could unify the two implementations, perhaps?